### PR TITLE
Jarema/async rewrite url

### PR DESCRIPTION
**Pull request by Jarema**

### CODE REVIEW

File: nats-experimental/src/lib.rs [hunk 1]
```diff
@@ -121,8 +124,18 @@ pub struct Connection {
 }
 
 impl Connection {
-    pub async fn connect(addr: impl ToSocketAddrs) -> Result<Connection, io::Error> {
-        let tcp_stream = TcpStream::connect(addr).await?;
+    pub async fn connect(addrs: impl IntoServerList) -> Result<Connection, io::Error> {
```
Comment (caspervonb) [inline nats-experimental/src/lib.rs:127]: Lets prefer explicit over implicit, unless we've made an explicit deliberate choice to do implicit conversion.
 ```suggestion
    pub async fn connect<A: ToServerAddrs>(addrs: A) -> Result<Connection, io::Error> {
```

File: nats-experimental/src/lib.rs [hunk 2]
```diff
@@ -436,9 +452,14 @@ impl Client {
...
     }
+
+    pub async fn flush(&mut self) -> Result<(), Error> {
+        self.sender.send(ClientOp::Flush).await?;
+        Ok(())
+    }
 }
 
-pub async fn connect<T: ToSocketAddrs>(addr: T) -> Result<Client, io::Error> {
+pub async fn connect(addr: impl IntoServerList) -> Result<Client, io::Error> {
```
Comment (caspervonb) [inline nats-experimental/src/lib.rs:462]: Same as above

File: nats-experimental/src/lib.rs [hunk 3]
```diff
@@ -502,3 +523,150 @@ impl Stream for Subscriber {
         self.receiver.poll_recv(cx)
     }
 }
+
+/// Address of a NATS server.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ServerAddress(Url);
```
Comment (caspervonb) [inline nats-experimental/src/lib.rs:529]:
```suggestion
pub struct ServerAddr(Url);
```

File: nats-experimental/src/lib.rs [hunk 4]
```diff
@@ -502,3 +523,150 @@ impl Stream for Subscriber {
...
+
+/// Capability to convert into a list of NATS server addresses.
+///
+/// There are several implementations ensuring the easy passing of one or more server addresses to
+/// functions like [`crate::connect()`].
+pub trait IntoServerList {
```
Comment (caspervonb) [inline nats-experimental/src/lib.rs:535]:
```suggestion
pub trait ToServerAddrs {
```

**Review by caspervonb — COMMENTED** [review]
Since we're bringing this over and can break it.
It would be good to align this a bit closer to prior art in terms of naming, e.g `SocketAddr`, `IpAddr`.

File: nats-experimental/src/lib.rs (hunk 1 above)
Comment (Jarema) [reply nats-experimental/src/lib.rs:127]: agree. Despite not liking abbreviations in Rust :). Fixef.

**Review by Jarema — COMMENTED** [review]
question :).

File: nats-experimental/src/lib.rs [hunk 5]
```diff
@@ -502,3 +523,143 @@ impl Stream for Subscriber {
...
+    fn to_server_addrs(&self) -> io::Result<Self::Iter> {
+        Ok(Some(self.clone()).into_iter())
+    }
+}
+
+impl ToServerAddrs for str {
+    type Iter = vec::IntoIter<ServerAddr>;
+    fn to_server_addrs(&self) -> io::Result<Self::Iter> {
+        self.parse::<ServerAddr>()
+            .map(|addr| vec![addr].into_iter())
```
[the remark this answers is not in the document]
Comment (caspervonb) [reply nats-experimental/src/lib.rs:641]: Left-over from when there was a comma split in here. We can use option into_iter instead.

**Comment by caspervonb** [comment]
Continued in https://github.com/nats-io/nats.rs/pull/350

**Pull request closed without merging**